pkg/caching/interceptors.go

```diff
@@ -0,0 +1,325 @@
+package caching
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/rancher/opni/pkg/storage"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/runtime/protoimpl"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+const (
+	// headers
+	CacheTypeClient = "client"
+	CacheTypeServer = "server"
+	// values
+
+	GrpcNoCache = "no-cache"
+	// server can also manually set this to revalidate the cache
+	GrpcMustRevalidateCache = "must-revalidate"
+	GrpcCacheMiss           = "miss"
+	GrpcCacheHit            = "hit"
+
+	GrpcCacheIdentifier = "x-cache-req-id"
+)
+
+var (
+	GrpcCacheControlHeader = func(cacheType string) string {
+		return "cache-control-" + cacheType
+	}
+	GrpcCacheStatusHeader = func(fullMethod string) string {
+		return "x-cache-" + fullMethod
+	}
+
+	GrpcMaxAge = func(d time.Duration) string {
+		return fmt.Sprintf("max-age=%d", int(d.Seconds()))
+	}
+	// maps proto message descriptor name to a function
+	// that returns a string that is not guaranteed to be unique,
+	// uniqueness is guaranteed when the partial key is appended to the full req method
+	knownCacheableTypes = map[string]func(msg protoreflect.ProtoMessage) string{}
+	knownTypesMu        sync.Mutex
+)
+
+func init() {
+	RegisterProtoType(&emptypb.Empty{}, func(_ protoreflect.ProtoMessage) string {
+		return "empty"
+	})
+}
+
+func protoId(msg protoreflect.ProtoMessage) string {
+	return string(msg.ProtoReflect().Descriptor().FullName().Name())
```

Comment:
Use the FullName here instead of the short name

Reply:
yep, good catch